wit/types.wit

```diff
@@ -0,0 +1,158 @@
+// The `wasi:http/types` interface is meant to be imported by components to
+// define the HTTP resource types and operations used by the component's
+// imported and exported interfaces.
+default interface types {
+  use io.streams.{input-stream, output-stream}
+  use poll.poll.{pollable}
+  
+  // This type corresponds to HTTP standard Methods.
+  variant method {
+    get,
+    head,
+    post,
+    put,
+    delete,
+    connect,
+    options,
+    trace,
+    patch,
+    other(string)
+  }
+
+  // This type corresponds to HTTP standard Related Schemes.
+  variant scheme {
+    HTTP,
+    HTTPS,
+    other(string)
+  }
+
+  // This type enumerates the different kinds of errors that may occur when
+  // initially returning a response.
+  variant error {
+      invalid-url(string),
+      timeout-error(string),
+      protocol-error(string),
+      status-error(u16),
+      unexpected-error(string)
+  }
+
+  // This following block defines the `fields` resource which corresponds to
+  // HTTP standard Fields. Soon, when resource types are added, the `type
+  // fields = u32` type alias can be replaced by a proper `resource fields`
+  // definition containing all the functions using the method syntactic sugar.
+  type fields = u32
+  drop-fields: func(fields: fields)
+  new-fields: func(entries: list<tuple<string,string>>) -> fields
+  fields-get: func(fields: fields, name: string) -> list<string>
+  fields-set: func(fields: fields, name: string, value: list<string>)
+  fields-delete: func(fields: fields, name: string)
+  fields-append: func(fields: fields, name: string, value: string)
+  fields-entries: func(fields: fields) -> list<tuple<string,string>>
+  fields-clone: func(fields: fields) -> fields
+
+  type headers = fields
+  type trailers = fields
+
+  // The following block defines the `body` type which corresponds to the HTTP
+  // standard Contents. With Preview3, all of these fields can be replaced by a
+  // single type definition:
+  //
+  //   type body = stream<u8, option<trailers>>
```

Comment:
I initially I was going to simply have a method returning a (pseudo) `future<trailers>` (as we discussed a while back), but then I realized that this introduces a tricky question: if I `poll_oneoff` the trailers without concurrently consuming the body stream, does that (1) deadlock (b/c we're not consuming the body, there is backpressure, so we never get the trailers), (2) implicitly discard/ignore the body (so that we can get to the trailers)?  Putting the trailers as the "final value" of the stream avoids this design problem.  But agreed that "`body`" is a misnomer if the type includes trailers.  What do you think of [this](https://github.com/WebAssembly/wasi-http/pull/3/commits/13db7f6e92184d9e4defefc9181043a5d0121b7d) alternative naming that still keeps the same API shape, but avoids calling the composite "body"?